src/pybind/mgr/cephadm/serve.py

```diff
@@ -633,9 +643,85 @@ def _apply_all_services(self) -> bool:
+            conflicts: List[orchestrator.DaemonDescription],
+            to_deploy: List[CephadmDaemonDeploySpec],
+            to_remove: List[orchestrator.DaemonDescription]
+        ) -> Tuple[bool, Set[str], List[str]]:
+            r: bool = False
+            removed_conflict_daemons, conflict_hosts_altered = await self.remove_given_daemons(conflicts)
+            if removed_conflict_daemons:
+                r = True
+                self.mgr.spec_store.mark_needs_configuration(spec.service_name())
+                hosts_altered.update(conflict_hosts_altered)
+            daemons_placed, daemon_deployed_hosts, daemon_place_fails = await self.deploy_given_daemons(to_deploy)
+            if daemons_placed:
+                r = True
+                hosts_altered.update(daemon_deployed_hosts)
+            removed_daemons, removed_daemon_hosts = await self.remove_given_daemons(to_remove)
+            if removed_daemons:
+                r = True
+                self.mgr.spec_store.mark_needs_configuration(spec.service_name())
+                hosts_altered.update(conflict_hosts_altered)
+            if self.mgr.spec_store.needs_configuration(spec.service_name()):
+                svc = service_registry.get_service(spec.service_type)
+                svc.config(spec)
+                self.mgr.spec_store.mark_configured(spec.service_name())
+            return (r, hosts_altered, daemon_place_fails)
+
+        async def _deploy_and_remove_all(
+            all_conflicts: List[orchestrator.DaemonDescription],
+            all_to_deploy: List[CephadmDaemonDeploySpec],
+            all_to_remove: List[orchestrator.DaemonDescription]
+        ) -> List[Tuple[bool, Set[str], List[str]]]:
+            futures = []
+
+            for host in self.mgr.cache.get_hosts():
+                conflicting_daemons = [dd for dd in all_conflicts if dd.hostname == host]
+                daemons_to_deploy = [dd for dd in all_to_deploy if dd.host == host]
+                daemons_to_remove = [dd for dd in all_to_remove if dd.hostname == host]
+                futures.append(_parallel_deploy_and_remove(host, conflicting_daemons, daemons_to_deploy, daemons_to_remove))
+
+            return await gather(*futures)
+
+        deploy_names = [d.name() for d in all_daemons_to_deploy]
+        rm_names = [d.name() for d in all_conflicting_daemons] + [d.name() for d in all_daemons_to_remove]
+        with self.mgr.async_timeout_handler(cmd=f'cephadm deploying ({deploy_names} and removing {rm_names} daemons)'):
+            results = self.mgr.wait_async(_deploy_and_remove_all(all_conflicting_daemons, all_daemons_to_deploy, all_daemons_to_remove))
+
+        if any(res[0] for res in results):
+            changed = True
+
+        altered_hosts: Set[str] = set()
+        for altered_host in [res[1] for res in results]:
+            altered_hosts.update(altered_host)
+
+        placement_failures: List[str] = []
+        for place_failures in [res[2] for res in results]:
```

Comment:
not for this first implementation, but I think we need named types for the result to avoid referring to res[x] and instead use more meaningful names.